homeassistant/components/braviatv/__init__.py

```diff
@@ -44,3 +67,237 @@ async def async_unload_entry(hass, config_entry):
+        self.program_media_type = None
+        self.audio_output = None
+        self.min_volume = None
+        self.max_volume = None
+        self.volume = None
+        self.is_on = False
+        # Assume that the TV is in Play mode
+        self.playing = True
+        self.state_lock = asyncio.Lock()
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=SCAN_INTERVAL,
+            request_refresh_debouncer=Debouncer(
+                hass, _LOGGER, cooldown=1.0, immediate=False
+            ),
+        )
+
+    def get_braviarc(self):
+        """Return BraviaRC instance."""
+        return self.braviarc
+
+    def _get_source(self):
+        """Return the name of the source."""
+        for key, value in self.content_mapping.items():
+            if value == self.content_uri:
+                return key
+
+    async def _async_refresh_volume(self):
+        """Refresh volume information."""
+        volume_info = await self.hass.async_add_executor_job(
+            self.braviarc.get_volume_info, self.audio_output
+        )
+        if volume_info is not None:
+            self.audio_output = volume_info.get("target")
+            self.volume = volume_info.get("volume")
+            self.min_volume = volume_info.get("minVolume")
+            self.max_volume = volume_info.get("maxVolume")
+            self.muted = volume_info.get("mute")
+            return True
+        return False
+
+    async def _async_refresh_channels(self):
+        """Refresh source and channels list."""
+        if not self.source_list:
+            self.content_mapping = await self.hass.async_add_executor_job(
+                self.braviarc.load_source_list
+            )
+            self.source_list = []
+            if not self.content_mapping:
+                return False
+            for key in self.content_mapping:
+                if key not in self.ignored_sources:
+                    self.source_list.append(key)
+        return True
+
+    async def _async_refresh_playing_info(self):
+        """Refresh playing information."""
+        playing_info = await self.hass.async_add_executor_job(
+            self.braviarc.get_playing_info
+        )
+        self.program_name = playing_info.get("programTitle")
+        self.channel_name = playing_info.get("title")
+        self.program_media_type = playing_info.get("programMediaType")
+        self.channel_number = playing_info.get("dispNum")
+        self.content_uri = playing_info.get("uri")
+        self.source = self._get_source()
+        self.duration = playing_info.get("durationSec")
+        self.start_date_time = playing_info.get("startDateTime")
+        if not playing_info:
+            self.channel_name = "App"
+
+    async def _async_update_data(self):
+        """Update TV info."""
+        if self.state_lock.locked():
+            return
+
+        power_status = await self.hass.async_add_executor_job(
```

Comment:
Please wrap all the sync work into a single function that we can schedule once on the executor. It's expensive to switch context between async and sync.